### PR TITLE
[Merged by Bors] - TY-3054 impl db code for source preferences

### PR DESCRIPTION
**Pull request by Robert-Steiner**
Ticket:
- [TY-3054]

[TY-3054]: https://xainag.atlassian.net/browse/TY-3054?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/core/src/storage/migrations/20220801152246_source_preference.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+--  but WITHOUT ANY WARRANTY; without even the implied warranty of
+--  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+--  GNU Affero General Public License for more details.
+--
+--  You should have received a copy of the GNU Affero General Public License
+--  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+CREATE TABLE IF NOT EXISTS SourcePreference (
+    source TEXT NOT NULL PRIMARY KEY,
+    -- 1 = preferred, 2 = excluded
```
Comment (Robert-Steiner) [inline discovery_engine_core/core/src/storage/migrations/20220801152246_source_preference.sql:17]: Do we want to keep the name `preferred` or would it be better to use `trusted`?

Comment (finiteprods) [reply discovery_engine_core/core/src/storage/migrations/20220801152246_source_preference.sql:17]: I'd say `trusted` as that's how we have been calling it

File: discovery_engine_core/core/src/storage/sqlite.rs [hunk 2]
```diff
@@ -1129,4 +1215,60 @@ mod tests {
...
+            .unwrap();
+
+        let trusted_db = storage.source_preference().fetch_trusted().await.unwrap();
+        assert_eq!(trusted_db, trusted_sources);
+        let excluded_db = storage.source_preference().fetch_excluded().await.unwrap();
+        assert_eq!(excluded_db, excluded_sources);
+
+        // set the excluded source and so far unknown source as new trusted sources
+        // excluded sources should be empty
+        // trusted sources should return {"d", "c"}
```
Comment (finiteprods) [inline discovery_engine_core/core/src/storage/sqlite.rs:1252]: :art:  would help to clarify a bit as in:
```suggestion
        // set the excluded source "c" and so far unknown source "d" as new trusted sources
        // excluded sources should be empty
        // trusted sources should return {"d", "c"}
```
:grey_question:  so when you set an excluded source as trusted, it stops being excluded. how about if you set a trusted source as excluded, is it worth testing this as well?

**Review by finiteprods — APPROVED** [review]
minor suggestion in the test otherwise looks good

File: discovery_engine_core/core/src/storage/migrations/20220801152246_source_preference.sql (hunk 1 above)
Comment (Robert-Steiner) [reply discovery_engine_core/core/src/storage/migrations/20220801152246_source_preference.sql]: i have changed it to `trusted`

File: discovery_engine_core/core/src/storage/sqlite.rs (hunk 2 above)
Comment (Robert-Steiner) [reply discovery_engine_core/core/src/storage/sqlite.rs]: good point 👍 i have added a test for this as well

**Comment by Robert-Steiner** [comment]
bors merge

**Pull request closed without merging**